doschema/jsvalidation.py

```diff
@@ -0,0 +1,232 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of DoSchema.
+# Copyright (C) 2016 CERN.
+#
+# DoSchema is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# DoSchema is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with DoSchema; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+import collections
+
+import jsonschema
+
+from doschema.errors import JSONSchemaCompatibilityException, \
+    JSONSchemaDependenciesException
+
+
+class JSONSchemaValidation(object):
+    """Class for checking compatibility between schemas."""
+
+    COLLECTION_KEYS = frozenset(['allOf', 'anyOf', 'oneOf'])
+
+    def __init__(self, ignore_index=True, resolver_factory=False):
+        """Constructor.
+
+        :param ignore_index: If set to True, which is default, it will ignore
+                             array indexes.
+        :param resolver: Resolver used to retrieve referenced schemas. If not
+                         provided it will use default.
+
+        """
+        self.ignore_index = ignore_index
+        self.fields_types_dict = {}
+        self.uri = None
+        self.resolver_factory = resolver_factory or jsonschema.RefResolver
+
+    def start(self, schema, uri):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param uri: Name of schema that is currently processed.
+        """
+        self.uri = uri
+        self.resolver = self.resolver_factory(base_uri=uri, referrer=schema)
+
+        self.traverse(schema, ())
```

Comment:
The root schema has to be of type `object`. There are multiple ways to check this, we could:
- separate `traverse` in multiple methods, and have a separate one for validating `objects`.
- add an optional argument to `traverse` which would check if the expected type and the found one match.
- return the found type from the `traverse` method and check it in `start`.